File: app/src/main/java/com/inspiringteam/xchange/di/AppComponent.java
package com.inspiringteam.xchange.di;

import android.app.Application;

import com.inspiringteam.xchange.data.source.RatesRepositoryModule;
import com.inspiringteam.xchange.di.scopes.AppScoped;

import dagger.BindsInstance;
import dagger.Component;
import dagger.android.AndroidInjector;
import dagger.android.support.AndroidSupportInjectionModule;

/**
 * This is the root Dagger component.
 * {@link AndroidSupportInjectionModule}
 * is the module from Dagger.Android that helps with the generation
 * and location of subcomponents, which will be in our case, activities
 */
@AppScoped
@Component(modules = {RatesRepositoryModule.class,
        AppModule.class,
        ActivityBindingModule.class,
        AndroidSupportInjectionModule.class})
public interface AppComponent extends AndroidInjector<com.inspiringteam.xchange.Application> {

    // we can now do DaggerAppComponent.builder().application(this).build().inject(this),
    // never having to instantiate any modules or say which module we are passing the application to.
    // Application will just be provided into our app graph

    @Component.Builder
    interface Builder {

        @BindsInstance
        AppComponent.Builder application(Application application);

        AppComponent build();
    }
}

File: app/src/main/java/com/inspiringteam/xchange/di/AppComponent.kt
package com.inspiringteam.xchange.di

import com.inspiringteam.xchange.data.source.QuakesRepositoryModule
import com.inspiringteam.xchange.di.scopes.AppScoped
import dagger.BindsInstance
import dagger.Component
import dagger.android.AndroidInjector
import dagger.android.support.AndroidSupportInjectionModule
import android.app.Application
import com.inspiringteam.xchange.App

/**
 * This is the root Dagger component.
 * [AndroidSupportInjectionModule]
 * is the module from Dagger.Android that helps with the generation
 * and location of subcomponents, which will be in our case, activities
 */
@AppScoped
@Component(
    modules = [
        QuakesRepositoryModule::class,
        ViewModelModule::class,
        UtilsModule::class,
        AppModule::class,
        ActivityBindingModule::class,
        AndroidSupportInjectionModule::class]
)
interface AppComponent : AndroidInjector<App> {

    // Application will just be provided into our app graph
    @Component.Builder
    interface Builder {

        @BindsInstance
        fun application(application: Application): AppComponent.Builder

        fun build(): AppComponent
    }
}